### PR TITLE
feat: add directml support

### PR DESCRIPTION
**Pull request by thewh1teagle**
Add support for directml execution provider
Improved version of https://github.com/k2-fsa/sherpa-onnx/pull/154

Also compared whisper medium execution time on CPU vs DirectML (amd ryzten 5 4500u, 20s audio duration)

CPU: 00:01:08.41
GPU: 00:00:37.21

The only issue in the bindings is that I had to copy the onnxruntime.dll file into the executable folder from `build/_deps/onnxruntime-src/native/x64`

In addition it doesn't utilize the full resources of the GPU. only 5%. Not sure how can we make it uses more.

### CODE REVIEW

File: cmake/onnxruntime-win-x64-directml.cmake [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+if(NOT onnxruntime_POPULATED)
+  message(STATUS "Downloading onnxruntime from ${onnxruntime_URL}")
+  FetchContent_Populate(onnxruntime)
+endif()
+message(STATUS "onnxruntime is downloaded to ${onnxruntime_SOURCE_DIR}")
+
+# for static libraries, we use onnxruntime_lib_files directly below
+include_directories(${onnxruntime_SOURCE_DIR}/build/native/include)
+
+file(GLOB onnxruntime_lib_files "${onnxruntime_SOURCE_DIR}/runtimes/win-x64/native/*.lib")
```
Comment (csukuangfj) [inline cmake/onnxruntime-win-x64-directml.cmake]: Please also glob `*.dll`

I suggest that you use
```
file(GLOB onnxruntime_lib_files "${onnxruntime_SOURCE_DIR}/runtimes/win-x64/native/onnxruntime.*")
```

**Review by csukuangfj — CHANGES_REQUESTED** [review]

**Comment by csukuangfj** [comment]
After unpacking the nuget package, I get
```
-rw-r--r-- 1 root root  12M Feb 25  2023 runtimes/win-arm64/native/onnxruntime.dll
-rw-r--r-- 1 root root 3.2K Feb 25  2023 runtimes/win-arm64/native/onnxruntime.lib
-rw-r--r-- 1 root root 8.9M Feb 25  2023 runtimes/win-arm/native/onnxruntime.dll
-rw-r--r-- 1 root root 3.2K Feb 25  2023 runtimes/win-arm/native/onnxruntime.lib
-rw-r--r-- 1 root root  11M Feb 25  2023 runtimes/win-x64/native/onnxruntime.dll
-rw-r--r-- 1 root root 3.2K Feb 25  2023 runtimes/win-x64/native/onnxruntime.lib
-rw-r--r-- 1 root root 9.6M Feb 25  2023 runtimes/win-x86/native/onnxruntime.dll
-rw-r--r-- 1 root root 3.2K Feb 25  2023 runtimes/win-x86/native/onnxruntime.lib
```

So the nuget package contains the shared library of onnxruntime.

---

> The only issue in the bindings is that I had to copy the onnxruntime.dll file into the executable folder from build/_deps/onnxruntime-src/native/x64

After fixing the code review comment, I think you should fix this problem.

File: cmake/onnxruntime-win-x64-directml.cmake [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+if(NOT CMAKE_SYSTEM_NAME STREQUAL Windows)
+  message(FATAL_ERROR "This file is for Windows only. Given: ${CMAKE_SYSTEM_NAME}")
+endif()
+
+if(NOT (CMAKE_VS_PLATFORM_NAME STREQUAL X64 OR CMAKE_VS_PLATFORM_NAME STREQUAL x64))
+  message(FATAL_ERROR "This file is for Windows x64 only. Given: ${CMAKE_VS_PLATFORM_NAME}")
+endif()
+
+if(BUILD_SHARED_LIBS)
```
Comment (csukuangfj) [inline cmake/onnxruntime-win-x64-directml.cmake]: Is it only for share libraries?

Comment (thewh1teagle) [reply cmake/onnxruntime-win-x64-directml.cmake]: Not sure. How can I find out? Should I omit that error statement?

**Comment by thewh1teagle** [comment]
> After fixing the code review comment, I think you should fix this problem.

Not sure what's changed, but suddenly it doesn't work with directml

```console
PS C:\Users\User\Documents\code\sherpa-rs2> $start = Get-Date; cargo run --features "directml" --example transcribe ; $elapsed = (Get-Date) - $start; Write-Output "Elapsed time: $elapsed"
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.25s
     Running `target\debug\examples\transcribe.exe`
C:\Users\User\Documents\code\sherpa-rs2\target\debug\build\sherpa-rs-sys-1fd886a7cbe68e85\out\sherpa-onnx\sherpa-onnx\c-api\c-api.cc:convertConfig:427 OfflineRecognizerConfig(feat_config=FeatureExtractorConfig(sampling_rate=16000, feature_dim=512, low_freq=20, high_freq=-400, dither=0), model_config=OfflineModelConfig(transducer=OfflineTransducerModelConfig(encoder_filename="", decoder_filename="", joiner_filename=""), paraformer=OfflineParaformerModelConfig(model=""), nemo_ctc=OfflineNemoEncDecCtcModelConfig(model=""), whisper=OfflineWhisperModelConfig(encoder="sherpa-onnx-whisper-tiny/tiny-encoder.onnx", decoder="sherpa-onnx-whisper-tiny/tiny-decoder.onnx", language="en", task="transcribe", tail_paddings=-1), tdnn=OfflineTdnnModelConfig(model=""), zipformer_ctc=OfflineZipformerCtcModelConfig(model=""), wenet_ctc=OfflineWenetCtcModelConfig(model=""), sense_voice=OfflineSenseVoiceModelConfig(model="", language="", use_itn=False), telespeech_ctc="", tokens="sherpa-onnx-whisper-tiny/tiny-tokens.txt", num_threads=2, debug=True, provider="directml", model_type="", modeling_unit="cjkchar", bpe_vocab=""), lm_config=OfflineLMConfig(model="", scale=1), ctc_fst_decoder_config=OfflineCtcFstDecoderConfig(graph="", max_active=3000), decoding_method="greedy_search", max_active_paths=4, hotwords_file="", hotwords_score=1.5, blank_penalty=0, rule_fsts="", rule_fars="")
C:\Users\User\Documents\code\sherpa-rs2\target\debug\build\sherpa-rs-sys-1fd886a7cbe68e85\out\sherpa-onnx\sherpa-onnx\csrc\offline-whisper-model.cc:InitEncoder:243 ---encoder---
n_audio_head=6
sot_prev=50361
blank_id=220
n_text_layer=4
version=1
model_type=whisper-tiny
maintainer=k2-fsa
sot_lm=50360
n_mels=80
no_timestamps=50363
all_language_tokens=50317,50296,50261,50276,50337,50260,50271,50279,50298,50319,50301,50272,50268,50352,50333,50322,50330,50347,50341,50303,50320,50267,50314,50344,50331,50321,50273,50350,50305,50357,50355,50286,50353,50309,50262,50270,50292,50312,50289,50325,50274,50285,50328,50284,50287,50311,50315,50283,50345,50346,50332,50338,50313,50354,50299,50295,50297,50281,50349,50327,50335,50306,50282,50326,50342,50351,50259,50348,50269,50264,50304,50291,50334,50275,50280,50340,50288,50310,50316,50323,50336,50324,50265,50308,50294,50356,50277,50302,50293,50278,50329,50343,50318,50290,50263,50266,50300,50339,50307
n_audio_layer=4
n_audio_state=384
n_audio_ctx=1500
n_vocab=51865
n_text_ctx=448
n_text_state=384
n_text_head=6
sot_sequence=50258,50259,50359
all_language_codes=sq,ml,de,hi,uz,zh,nl,he,sk,gl,lv,ar,tr,haw,gu,si,be,bo,tk,sr,mr,pt,mn,sa,tg,pa,sv,as,sl,su,ba,hu,ln,br,es,ca,bg,hy,th,yo,it,da,oc,ro,ta,is,bs,cs,lb,my,sd,fo,ne,ha,te,mi,cy,el,mg,af,yi,kn,ms,so,nn,tt,en,tl,pl,ko,az,hr,am,id,uk,ps,no,eu,kk,km,lo,sn,fr,mk,la,jw,fi,bn,lt,vi,ka,mt,sw,ur,ru,ja,fa,ht,et
sot=50258
sot_index=0
eot=50257
is_multilingual=1
no_speech=50362
non_speech_tokens=1,2,7,8,9,10,14,25,26,27,28,29,31,58,59,60,61,62,63,90,91,92,93,359,503,522,542,873,893,902,918,922,931,1350,1853,1982,2460,2627,3246,3253,3268,3536,3846,3961,4183,4667,6585,6647,7273,9061,9383,10428,10929,11938,12033,12331,12562,13793,14157,14635,15265,15618,16553,16604,18362,18956,20075,21675,22520,26130,26161,26435,28279,29464,31650,32302,32470,36865,42863,47425,49870,50254
transcribe=50359
translate=50358


2024-07-20 22:12:00.9327135 [E:onnxruntime:, sequential_executor.cc:494 onnxruntime::ExecuteKernel] Non-zero status code returned while running Gelu node. Name:'Gelu' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(2340)\onnxruntime.dll!00007FFD1794E70A: (caller: 00007FFD1794DC5B) Exception(3) tid(3554) 80070057 The parameter is incorrect.

C:\Users\User\Documents\code\sherpa-rs2\target\debug\build\sherpa-rs-sys-1fd886a7cbe68e85\out\sherpa-onnx\sherpa-onnx/csrc/offline-recognizer-whisper-impl.h:DecodeStream:176

Caught exception:

Non-zero status code returned while running Gelu node. Name:'Gelu' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(2340)\onnxruntime.dll!00007FFD1794E70A: (caller: 00007FFD1794DC5B) Exception(3) tid(3554) 80070057 The parameter is incorrect.


Return an empty result. Number of input frames: 2000, Current tail paddings: 1000. If you see a lot of such exceptions, please consider using a larger --whisper-tail-paddings
WhisperRecognizerResult { text: "" }
Elapsed time: 00:00:03.4868930
```

**Comment by csukuangfj** [comment]
please try a new wav.

**Comment by csukuangfj** [comment]
please try a new wav.

**Comment by thewh1teagle** [comment]
> please try a new wav.

Didn't helped. maybe the model doesn't support DirectML? I was pretty sure it already worked but maybe I didn't noticed the error from the first place

**Comment by csukuangfj** [comment]
> After unpacking the nuget package, I get
> 
> ```
> -rw-r--r-- 1 root root  12M Feb 25  2023 runtimes/win-arm64/native/onnxruntime.dll
> -rw-r--r-- 1 root root 3.2K Feb 25  2023 runtimes/win-arm64/native/onnxruntime.lib
> -rw-r--r-- 1 root root 8.9M Feb 25  2023 runtimes/win-arm/native/onnxruntime.dll
> -rw-r--r-- 1 root root 3.2K Feb 25  2023 runtimes/win-arm/native/onnxruntime.lib
> -rw-r--r-- 1 root root  11M Feb 25  2023 runtimes/win-x64/native/onnxruntime.dll
> -rw-r--r-- 1 root root 3.2K Feb 25  2023 runtimes/win-x64/native/onnxruntime.lib
> -rw-r--r-- 1 root root 9.6M Feb 25  2023 runtimes/win-x86/native/onnxruntime.dll
> -rw-r--r-- 1 root root 3.2K Feb 25  2023 runtimes/win-x86/native/onnxruntime.lib
> ```
> 
> So the nuget package contains the shared library of onnxruntime.
> 
> > The only issue in the bindings is that I had to copy the onnxruntime.dll file into the executable folder from build/_deps/onnxruntime-src/native/x64
> 
> After fixing the code review comment, I think you should fix this problem.

Please have a look at this comment.

You can refer to how we support windows x64 GPU with CUDA.

The directML package provides only shared libraries.

**Comment by csukuangfj** [comment]
> > Please have a look at this comment.
> > You can refer to how we support windows x64 GPU with CUDA.
> > The directML package provides only shared libraries.
> 
> Added new commits. let me know if something is missing. Anyway I get the same error with whisper example.
> 
> Update: Added commit that fixed the issue. The issue was that onnxruntime.dll used the older system DirectML.dll file. I added new one in Cmake. Now it works with DirectML with tiny whisper model. But with medium model I get the error:
> 
> error.log
> ```
> PS C:\Users\User\Documents\code\sherpa-onnx> .\build\bin\Release\offline-stt-c-api.exe
> C:\Users\User\Documents\code\sherpa-onnx\sherpa-onnx\c-api\c-api.cc:convertConfig:427 OfflineRecognizerConfig(feat_config=FeatureExtractorConfig(sampling_rate=16000, feature_dim=512, low_freq=20, high_freq=-400, dither=0), model_config=OfflineModelConfig(transducer=OfflineTransducerModelConfig(encoder_filename="", decoder_filename="", joiner_filename=""), paraformer=OfflineParaformerModelConfig(model=""), nemo_ctc=OfflineNemoEncDecCtcModelConfig(model=""), whisper=OfflineWhisperModelConfig(encoder="sherpa-onnx-whisper-medium/medium-encoder.onnx", decoder="sherpa-onnx-whisper-medium/medium-decoder.onnx", language="en", task="transcribe", tail_paddings=-1), tdnn=OfflineTdnnModelConfig(model=""), zipformer_ctc=OfflineZipformerCtcModelConfig(model=""), wenet_ctc=OfflineWenetCtcModelConfig(model=""), sense_voice=OfflineSenseVoiceModelConfig(model="", language="", use_itn=False), telespeech_ctc="", tokens="sherpa-onnx-whisper-medium/medium-tokens.txt", num_threads=1, debug=True, provider="directml", model_type="", modeling_unit="cjkchar", bpe_vocab=""), lm_config=OfflineLMConfig(model="", scale=1), ctc_fst_decoder_config=OfflineCtcFstDecoderConfig(graph="", max_active=3000), decoding_method="greedy_search", max_active_paths=4, hotwords_file="", hotwords_score=1.5, blank_penalty=0, rule_fsts="", rule_fars="")
> C:\Users\User\Documents\code\sherpa-onnx\sherpa-onnx\csrc\offline-whisper-model.cc:InitEncoder:243 ---encoder---
> n_audio_head=16
> sot_prev=50361
> blank_id=220
> n_text_layer=24
> version=1
> model_type=whisper-medium
> maintainer=k2-fsa
> sot_lm=50360
> n_mels=80
> no_timestamps=50363
> all_language_tokens=50272,50341,50314,50335,50313,50317,50264,50326,50324,50337,50320,50318,50352,50355,50350,50263,50348,50275,50342,50349,50279,50331,50268,50287,50303,50340,50325,50297,50299,50293,50319,50329,50315,50280,50300,50298,50306,50276,50286,50277,50281,50336,50292,50328,50288,50302,50310,50269,50312,50270,50333,50338,50262,50285,50354,50307,50284,50323,50346,50301,50321,50339,50334,50265,50309,50291,50266,50294,50267,50271,50308,50260,50305,50353,50289,50356,50282,50304,50332,50283,50274,50311,50357,50296,50259,50290,50278,50343,50322,50261,50351,50347,50327,50345,50330,50273,50295,50344,50316
> n_audio_layer=24
> n_audio_state=1024
> n_audio_ctx=1500
> n_vocab=51865
> n_text_ctx=448
> n_text_state=1024
> n_text_head=16
> sot_sequence=50258,50259,50359
> all_language_codes=ar,tk,mn,yi,ne,sq,ko,so,sn,uz,mr,sw,haw,ba,as,ru,tl,id,nn,mg,he,tg,tr,ta,sr,ps,yo,cy,te,lt,gl,ka,bs,uk,fa,sk,kn,hi,hu,fi,el,lo,bg,oc,no,bn,eu,pl,hy,ca,gu,fo,es,da,ha,et,ro,km,my,lv,pa,ht,am,fr,br,hr,ja,la,pt,nl,mk,zh,sl,ln,th,jw,ms,az,sd,cs,it,is,su,ml,en,ur,vi,mt,si,de,tt,bo,af,lb,be,sv,mi,sa,kk
> sot=50258
> sot_index=0
> eot=50257
> is_multilingual=1
> no_speech=50362
> non_speech_tokens=1,2,7,8,9,10,14,25,26,27,28,29,31,58,59,60,61,62,63,90,91,92,93,359,503,522,542,873,893,902,918,922,931,1350,1853,1982,2460,2627,3246,3253,3268,3536,3846,3961,4183,4667,6585,6647,7273,9061,9383,10428,10929,11938,12033,12331,12562,13793,14157,14635,15265,15618,16553,16604,18362,18956,20075,21675,22520,26130,26161,26435,28279,29464,31650,32302,32470,36865,42863,47425,49870,50254
> transcribe=50359
> translate=50358
> 
> 
> 2024-07-22 00:04:32.5505644 [E:onnxruntime:, sequential_executor.cc:494 onnxruntime::ExecuteKernel] Non-zero status code returned while running MemcpyToHost node. Name:'Memcpy_token_244' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(2429)\onnxruntime.dll!00007FFFD424E1D6: (caller: 00007FFFD4775FC7) Exception(3) tid(22fc) 887A0006 The GPU will not respond to more commands, most likely because of an invalid command passed by the calling application.
> 
> C:\Users\User\Documents\code\sherpa-onnx\sherpa-onnx/csrc/offline-recognizer-whisper-impl.h:DecodeStream:176 
> 
> Caught exception:
> 
> Non-zero status code returned while running MemcpyToHost node. Name:'Memcpy_token_244' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(2429)\onnxruntime.dll!00007FFFD424E1D6: (caller: 00007FFFD4775FC7) Exception(3) tid(22fc) 887A0006 The GPU will not respond to more commands, most likely because of an invalid command passed by the calling application.
> 
> 
> Return an empty result. Number of input frames: 2000, Current tail paddings: 1000. If you see a lot of such exceptions, please consider using a larger --whisper-tail-paddings
> Decoded text:
> ```

I will try to merge
https://github.com/k2-fsa/sherpa-onnx/pull/1037
and add C++ code for it.

**Review by csukuangfj — APPROVED** [review]
Thank you for your contribution!

**Comment by thewh1teagle** [comment]
> I will try to merge
> #1037
> and add C++ code for it.

I don't think it's related to that.
It works with tiny model. see https://github.com/k2-fsa/sherpa-onnx/issues/1240

**Comment by HaujetZhao** [comment]
请教一个问题。我下载了 sherpa-onnx-master 源码，然后在其中安装支持 directml 的，用了如下 powershell 命令：

```powershell
$env:SHERPA_ONNX_CMAKE_ARGS="-A x64 -DSHERPA_ONNX_ENABLE_DIRECTML=ON -DCMAKE_BUILD_TYPE=Release"
pip install . --no-cache-dir
```

安装好之后，provider 使用 directml ，提示：

```
D:\sherpa-onnx-master\sherpa-onnx\csrc\session.cc:GetSessionOptionsImpl:198 Failed to enable DirectML: D:\a\_work\1\s\onnxruntime\core\providers\dml\dml_provider_factory.cc(163)\onnxruntime.dll!00007FFA6B50C842: (caller: 00007FFA6B50C61E) Exception(1) tid(6708) 887A0004 指定的设备接口或功能级别在此系统上不受支持。
```

而我安装的 onnxruntime-directml 库是能正常用 directml 的。我是独显笔记本，有集显和5050独显。

我把 `session.cc` 中的这部分

```
#if defined(_WIN32) && SHERPA_ONNX_ENABLE_DIRECTML == 1
      sess_opts.DisableMemPattern();
      sess_opts.SetExecutionMode(ORT_SEQUENTIAL);
      int32_t device_id = 0;
      OrtStatus *status =
          OrtSessionOptionsAppendExecutionProvider_DML(sess_opts, device_id);
      if (status) {
        const auto &api = Ort::GetApi();
        const char *msg = api.GetErrorMessage(status);
        SHERPA_ONNX_LOGE("Failed to enable DirectML: %s. Fallback to cpu", msg);
        api.ReleaseStatus(status);
      }
#else
```

device_id 改成 1 再重新编译安装，仍不能用。

我安装 sherpa-onnx 的方式有问题吗？用 ai 搞了半天了，也搞不定。


---

我用下面的脚本可以验证 onnxruntime-directml 能正常运行：

```python
#!/usr/bin/env python3
"""
创建大模型并加载到 GPU 显存,用于监控 GPU 使用情况
"""

import numpy as np
import onnxruntime as ort
import onnx
from onnx import helper, TensorProto
import time

print("=" * 60)
print("DirectML GPU 显存占用测试")
print("=" * 60)

# 创建一个较大的模型 (约 500MB)
print("\n创建大模型 (约 500MB)...")

# 创建一个大的卷积网络
input_tensor = helper.make_tensor_value_info('input', TensorProto.FLOAT, [1, 3, 1024, 1024])
output_tensor = helper.make_tensor_value_info('output', TensorProto.FLOAT, [1, 64, 1024, 1024])

# 创建卷积权重 (3x64x3x3 = 1728 个参数)
conv_weight = np.random.randn(64, 3, 3, 3).astype(np.float32)
conv_weight_tensor = helper.make_tensor(
    'conv_weight',
    TensorProto.FLOAT,
    [64, 3, 3, 3],
    conv_weight.flatten().tolist()
)

# 创建卷积节点
conv_node = helper.make_node(
    'Conv',
    inputs=['input', 'conv_weight'],
    outputs=['output'],
    kernel_shape=[3, 3],
    pads=[1, 1, 1, 1]
)

# 创建图
graph = helper.make_graph(
    [conv_node],
    'large_model',
    [input_tensor],
    [output_tensor],
    [conv_weight_tensor]
)

# 创建模型
model = helper.make_model(graph, producer_name='test', opset_imports=[helper.make_opsetid("", 11)])
model.ir_version = 8

# 保存模型
model_path = 'large_test_model.onnx'
onnx.save(model, model_path)
print(f"✓ 模型已创建: {model_path}")

# 测试两个 GPU
for device_id in [0, 1]:
    print(f"\n{'=' * 60}")
    print(f"测试 GPU {device_id}")
    print(f"{'=' * 60}")
    
    try:
        # 创建 DirectML session
        print(f"\n正在加载模型到 GPU {device_id}...")
        dml_options = {'device_id': device_id}
        
        sess = ort.InferenceSession(
            model_path,
            providers=[
                ('DmlExecutionProvider', dml_options),
                'CPUExecutionProvider'
            ]
        )
        print(f"✓ 模型已加载到 GPU {device_id}")
        print(f"使用的 Provider: {sess.get_providers()}")
        
        # 创建大的输入数据 (约 12MB)
        print(f"\n创建输入数据...")
        input_data = np.random.randn(1, 3, 1024, 1024).astype(np.float32)
        print(f"✓ 输入数据大小: {input_data.nbytes / 1024 / 1024:.1f} MB")
        
        # 运行推理,将数据加载到 GPU
        print(f"\n运行推理,将数据加载到 GPU {device_id} 显存...")
        output = sess.run(None, {'input': input_data})
        print(f"✓ 推理完成")
        print(f"输出数据大小: {output[0].nbytes / 1024 / 1024:.1f} MB")
        
        # 暂停,让用户查看 GPU 使用情况
        print(f"\n" + "=" * 60)
        print(f"模型和数据已加载到 GPU {device_id} 显存!")
        print(f"=" * 60)
        print(f"\n请打开任务管理器查看 GPU 使用情况:")
        print(f"  1. 按 Ctrl+Shift+Esc 打开任务管理器")
        print(f"  2. 切换到 '性能' 标签")
        print(f"  3. 查看 GPU {device_id} 的显存使用情况")
        print(f"\nGPU 0 = Intel Arc 140T")
        print(f"GPU 1 = NVIDIA RTX 5050")
        print(f"\n按 Enter 继续测试下一个 GPU...")
        input()
        
        # 清理
        del sess
        del input_data
        del output
        print(f"\n✓ GPU {device_id} 资源已释放\n")
        
    except Exception as e:
        print(f"\n✗ GPU {device_id} 测试失败: {e}")
        import traceback
        traceback.print_exc()
        print(f"\n按 Enter 继续...")
        input()

# 清理模型文件
import os
try:
    os.remove(model_path)
    print(f"\n✓ 已清理测试文件: {model_path}")
except:
    pass

print(f"\n{'=' * 60}")
print("测试完成!")
print(f"{'=' * 60}\n")
```

**Pull request merged**